accepted/2025/nuget-mcp.md

```diff
@@ -0,0 +1,234 @@
+
+Some Microsoft MCP servers are implemented in .NET but distributed via npm, for example [@azure/mcp](https://www.npmjs.com/package/@azure/mcp). This is a fine approach if a Node.js runtime and npm packaging are not concerns. For MCP authors that want to target an environment where the .NET runtime is available, and not depend on npm, Python, or Docker, we should enable an end-to-end experience using just .NET and the MCP client IDE of choice (e.g., VS Code).
+
+## Explanation
+
+### Improve MCP server authoring
+
+#### Today's experience
+
+To create a .NET MCP server today, the experience is somewhat self-guided and results in a NuGet package that looks no different from any other .NET CLI tool.
+
+An MCP server author can create a .NET console app, consume the [ModelContextProtocol NuGet package](https://www.nuget.org/packages/ModelContextProtocol), and implement tools that will be available to the LLM. The SDK helps with the stdio-based protocol that allows the end user's IDE to launch, discover, and invoke tools.
+
+The author will decide what CLI arguments or environment variables are needed to invoke the MCP server.
+
+When the MCP server is ready, the author will pack the [project as a tool with `<PackAsTool>`](https://learn.microsoft.com/en-us/dotnet/core/tools/global-tools-how-to-create) and publish the NuGet package to NuGet.org (or their private package feed), so that it can be used by other people.
+
+This approach works, but has a couple of problems:
+
+- The package has no indication that it is an MCP server. It just has the `DotnetTool` package type like any other .NET tool.
+  - NuGet.org and other package sources have no way of filtering packages to "MCP" or tailoring the browsing experience.
+- The package provides no guidance to the end user on how to start the MCP server.
+  - The end user doesn't have a good way of knowing the needed `command`, `args`, and `env` settings they should use.
+- The package author needs to set up their project just right, with a console app, .NET tool, and MCP dependency.
+
+#### New package type
+
+To address the filtering and discoverability problem, we will introduce an additional NuGet package type of `McpServer`.
+
+Currently, .NET tools get the `DotnetTool` package type, defined in the resulting package manifest (.nuspec file) generated by `dotnet pack`. [As of .NET 10](https://github.com/dotnet/sdk/pull/48039), it is possible to add additional package types to .NET tool packages using the `<PackageType>McpServer</PackageType>` MSBuild property in the tool .csproj.
+
+The resulting MCP server .nupkg will have two package types: `DotnetTool` and `McpServer`. For more information on package types, [see NuGet documentation](https://learn.microsoft.com/en-us/nuget/reference/nuspec#packagetypes).
+
+#### Startup instructions
```

Comment:
We should emphasize the `dotnet new mcpserver` or similar template that will have this already there.